libuavcan/include/libuavcan/media/interfaces.hpp

```diff
@@ -0,0 +1,255 @@
+ *                                  on a redundant set of interfaces to a single bus).
+ */
+template <typename InterfaceT, typename InterfacePtrT, std::size_t MaxSelectInterfacesVal = 2>
+class LIBUAVCAN_EXPORT InterfaceManager
+{
+public:
+    virtual ~InterfaceManager() = default;
+
+    /**
+     * The media-specific interface managed by this object.
+     */
+    using InterfaceType = InterfaceT;
+
+    /**
+     * The pointer type used for pointers to InterfaceType. This can be a raw pointer or a smart pointer.
+     * The temporization is provided to allow either based on how a given media layer is implemented.
+     */
+    using InterfacePtrType = InterfacePtrT;
+
+    /**
+     * The maximum number of interfaces the select method must support in a single invocation.
+     */
+    static constexpr std::size_t MaxSelectInterfaces = MaxSelectInterfacesVal;
+
+    /**
+     * Opens an interface for receiveing and transmitting.
+     *
+     * @param       interface_index         The index of the interface to open. See getHardwareInterfaceCount()
+     *                                      for the valid range of interface indicies. The behaviour of this method
+     *                                      given an index >= getHardwareInterfaceCount() is implementation defined.
+     * @param       filter_config           An array of frame filtering parameters. The contents and behaviour of
+     *                                      filters is dependant on the interface in use and the Frame type.
+     * @param       filter_config_length    The number of filter configurations in the filter_config array.
+     * @param[out]  out_interface           If successful, the pointer is set to an open interface abstraction
+     *                                      owned by this manager instance. This memory shall remain valid while
+     *                                      the manager object is valid. It is undefined behavior to destroy a manager
+     *                                      with open interfaces still allocated.
+     *                                      Implementations must define the semantics for calling this method multiple
+     *                                      times with the same interface_index and for calling closeInterface on
+     *                                      a shared pointer to an interface.
+     * @return libuavcan::Result::success if the interface was successfully opened and returned,
+     */
+    virtual libuavcan::Result openInterface(std::uint_fast8_t                                interface_index,
+                                            const typename InterfaceType::FrameType::Filter* filter_config,
+                                            std::size_t                                      filter_config_length,
+                                            InterfacePtrT&                                   out_interface) = 0;
```

Comment:
- Why does the library need to be involved in the interface life cycle management at all? Is it not an over-specification of the logic when you're enforcing a particular approach to managing the whole life cycle from creation to destruction of an interface instance? If the scope of the library is that of managing communication at the application level, then would it not be sufficient to demand that by the time the node class is first instantiated the required interfaces are already available, configured, and open, ready to be used, and that they remain this way at least until the node is destroyed? That should allow us to leave their lifecycle control entirely to the application- or interface-specific code.

- Can't configure the acceptance filters unless you know which subscriptions the application is going to require. Those, in turn, may be changing dynamically (probably not in a deeply embedded safety-critical application though, but it is possible), or they may become known only AFTER the application has finished its initialization. To support both cases I suggest having a dedicated method for configuring the acceptance filters.

Reply:
I'm realizing that I do need an "interface group" abstraction in the media layer after all. Let me add that, remove lifecycle management from the media APIs, and move select and filter config into the group interface since these are operations we want to be the same across all groups.

Reply:
As for the filters and lifecycles, what if there is a deeply embedded node where the filters had to be configured once and only once at initialization time (e.g. after the first write some peripheral registers become read-only and the system can only change filters again after a reset). Because of this, and because UAVCAN itself doesn't need to change the filters dynamically, we should define the filters at the same time we create an interface group. This allows implementations to do both interface initialization and filter configuration as part of the `createInterfaceGroup` method (that I am working on right now) or they can choose to expose more dynamic behavior in their specializations and implement the `createInterfaceGroup` method as a getter that also configures filters on existing interfaces.

Reply:
Does the transport layer actually need access to individual interfaces? It seems we need to know some Identifier unique to the interface a frame was received on but I think we always send messages out on all configured interfaces when transmitting. If so I can simplify this design changing `Interface` into `InterfaceGroup`. The transport layer then won't need to be aware of interfaces indices other than using an opaque identifier for interface on received frames.

Reply:
As for lifecycle: as I refactor this PR to include an `InterfaceGroup` I do think the manager should continue to have lifecycle methods. In this case the `createInterfaceGroup` and `releaseInterfaceGroup`. While release will be largely a noop on an MCU, for platforms like Linux this will encourage orderly cleanup of resources when the libuavcan stack is shutdown. The transport layer would invoke the manager's release after being "shut down" itself which ensures it can flush queues and release any memory it may have acquired before exiting. This is particularly important (whether one thinks such a design is a good or not) where an application is a monolithic process with many threads as tasks.

The create method is important for all platforms for reasons I mentioned above and to allow the application to provide the `InterfaceManager` object to the transport layer as part of a nothrow constructor allowing error handling on the create actions to happen later.

Reply:
These are sensible arguments. There is one caveat that needs mentioning in the docs though:

> While applications may expose APIs allowing dynamic modification of filters libuavcan does not require this and will only ever request filtering once per call to this method.

I think it should be said very explicitly here, in big letters perhaps, that if the implementation does not expose the dynamic filter reconfiguration API, the application will be unable to add new subscribers after the interface group is initialized.

Are we planning to implement the acceptance filter optimization algorithm as in the old library? If not, are we going to require the user to compute filter configurations manually?

I have implemented it in PyUAVCAN v1.0 (well, it's not 1.0 but we seem to have developed a weird versioning jargon here); the filters are reconfigured automatically whenever the application creates or terminates a subscription. Such a dynamic approach is certainly inappropriate for libuavcan, but it might encourage good filter usage if the user does not have to tinker with bit masks whenever a subscription set is changed. https://pyuavcan.readthedocs.io/en/latest/.apidoc_generated/pyuavcan.transport.can.media.html#pyuavcan.transport.can.media.optimize_filter_configurations

Reply:
Hmm. Perhaps an optional API is in order then? We could document that this configure API is allowed to return `libuavcan::Result::NotSupported` which would cascade up to failures in subscription APIs if called after the media layer was initialized?

Reply:
As I said I don't think an implicit automatic reconfiguration is an acceptable approach for libuavcan. For the sake of clearer resource allocation (CPU time in this case), it might be preferable to make the user trigger reconfiguration of acceptance filters manually (filter optimization is an expensive process, so some highly static applications may prefer to compute filters before run time, possibly storing multiple predefined configurations and switching between them as necessary).

If this is accepted, there will be no need to float up a NotSupported error up the subscription stack because the process of establishing a new subscription would be separated from the process of filter reinitialization. For this interface, it means that there should probably be a virtual method for filter reconfiguration that returns NotImplemented by default, which is invoked by the application's request.

Reply:
Okay. Just added this.